### PR TITLE
Handle mutable default arguments cleanly

### PR DESCRIPTION
**Pull request by abn**
When generating code, ensure that default list/dict arguments are initialised in local scope if unspecified or `None`.

The following diff shows the difference in the generated source file from `service.proto`.
```diff
5c5
< from typing import AsyncIterable, AsyncIterator, Iterable, List, Union
---
> from typing import AsyncIterable, AsyncIterator, Iterable, List, Optional, Union
35c35
<         self, *, name: str = "", comments: List[str] = []
---
>         self, *, name: str = "", comments: Optional[List[str]] = None
36a37
>         comments = comments or []
```

Due to the nature of the issue, the included test simply verifies that the default list instance initialised is not the same object used in any subsequent calls. This is done so by intercepting the calls made to the underlying `ServiceStub._unary_unary` method and verifying that the `comments` instance is not reusing the object.

### CODE REVIEW

**Comment by nat-n** [comment]
Nice. Maybe worth having a test case for this? I assume you've verified it manually?

**Comment by abn** [comment]
@nat-n I did test it manually. Although turns out I did not handle the edge case when `Optional` was never imported. Had to change tact to modify the code instead of the template. Added a test, although had to make use of `pytest-mock` since the affected code is in the generated files. This should now be good to go.

File: betterproto/plugin.py [hunk 1]
```diff
@@ -314,6 +314,24 @@ def generate_code(request, response):
...
+                        for f in input_message.get("properties"):
+                            if f["zero"] != "None" and (
+                                f["type"].startswith("List[")
+                                or f["type"].startswith("Dict[")
+                            ):
+                                output["typing_imports"].add("Optional")
+                                mutable_default_args.append((f["py_name"], f["zero"]))
+                                f["zero"] = "None"
+                            properties.append(f)
+                        input_message["properties"] = properties
```
Comment (boukeversteegh) [inline betterproto/plugin.py:333]: I'm a bit nervous to see the entire `properties` field being overridden, as it makes it not immediately clear if it is changed fundamentally or not. Since the values are dictionaries (objects), it should be possible to just modify them, instead of recreating the list?

Comment (abn) [reply betterproto/plugin.py]: Good call. Fixed it.

File: betterproto/tests/grpc/test_grpclib_client.py [hunk 2]
```diff
@@ -1,14 +1,18 @@
 import asyncio
+import sys
+
+import grpclib
+import grpclib.metadata
```
Comment (nat-n) [inline betterproto/tests/grpc/test_grpclib_client.py:5]: redundant import?

**Comment by boukeversteegh** [comment]
Ah, sorry it took me so long, but i finally get it.
The diff in the example was just a tad too short to get the context..

So:

- This is about the arguments of methods of ServiceStubs (aka Clients)
- When a Service accepts a message that contains a `repeated` field, a parameter was created with a default `[]`
- If the caller does not specify the parameter, the default parameter is used
- If that parameter is then modified by the ServiceStub, it will contain the modifications upon the next call.
- This would cause unexpected and hard to debug behavior

Correct?

What I still don't understand is:

- In what way might the default parameter accidentally become modified, and leak into the next call?

From what I see, the request object created on the fly, and then passed to grpcio for serialization and sending. In theory it should not ever be modified.

**Comment by abn** [comment]
@boukeversteegh that is correct. The change is to ensure generated code adheres to best practices. Since the project aims to generate clean pythonic code at best effort, I think this is a  change in that direction. 

I am not across all possible scenarios, present and future, of how this code gets used to say for certain that the default value (initialised at definition) will never be mutated prior to serialisation. This is more a defense-in-depth change.

File: betterproto/tests/grpc/test_grpclib_client.py (hunk 2 above)
Comment (abn) [reply betterproto/tests/grpc/test_grpclib_client.py:5]: Intentional. Without it, inspection will complain. Think `__all__` is defined in `grpclib/__init__.py`.

Comment (nat-n) [reply betterproto/tests/grpc/test_grpclib_client.py:5]: right, though I didn't see anything using grpclib.metadata specifically

Comment (abn) [reply betterproto/tests/grpc/test_grpclib_client.py:5]: It is used by existing test cases. See L59 for example below.

https://github.com/danielgtaylor/python-betterproto/blob/0cf8aaab6bddbdebcb0602b4ca36428d4cf70a5a/betterproto/tests/grpc/test_grpclib_client.py#L56-L124

**Comment by abn** [comment]
@boukeversteegh this should be ready for review again. The recent refactor in master makes the implementation here a bit cleaner.

**Comment by abn** [comment]
@boukeversteegh @nat-n rebased and ready again; had to rebased due to improt statement collision.
